plugins/corsair/fu-corsair-device.c

```diff
@@ -8,103 +8,31 @@
 
 #define CORSAIR_DEFAULT_VENDOR_INTERFACE_ID 1
-#define CORSAIR_ACTIVATION_TIMEOUT	    30000
-#define CORSAIR_MODE_BOOTLOADER		    3
-#define CORSAIR_FIRST_CHUNK_HEADER_SIZE	    7
-#define CORSAIR_NEXT_CHUNKS_HEADER_SIZE	    3
-#define CORSAIR_MAX_CMD_SIZE		    1024
-#define CORSAIR_TRANSACTION_TIMEOUT	    2000
-
-#define CORSAIR_OFFSET_CMD_VERSION	 0x03
-#define CORSAIR_OFFSET_CMD_CRC		 0x08
-#define CORSAIR_OFFSET_CMD_MODE		 0x03
-#define CORSAIR_OFFSET_CMD_STATUS	 0x02
-#define CORSAIR_OFFSET_CMD_FIRMWARE_SIZE 0x03
-#define CORSAIR_OFFSET_CMD_SET_MODE	 0x04
+
+#define CORSAIR_TRANSACTION_TIMEOUT	    4000
+#define CORSAIR_SUBDEVICE_POLL_PERIOD	    30000
+#define CORSAIR_SUBDEVICE_REBOOT_DELAY	    (4 * G_USEC_PER_SEC)
+#define CORSAIR_SUBDEVICE_RECONNECT_RETRIES 30
+#define CORSAIR_SUBDEVICE_RECONNECT_PERIOD  1000
 
 struct _FuCorsairDevice {
 	FuUsbDevice parent_instance;
 	FuCorsairDeviceKind device_kind;
 	guint8 vendor_interface;
-	guint8 epin;
-	guint8 epout;
-	guint16 cmd_write_size;
-	guint16 cmd_read_size;
+	gchar *subdeviceId;
+	FuCorsairBp *bp;
 };
 G_DEFINE_TYPE(FuCorsairDevice, fu_corsair_device, FU_TYPE_USB_DEVICE)
 
-typedef enum {
-	FU_CORSAIR_DEVICE_MODE_APPLICATION = 1,
-	FU_CORSAIR_DEVICE_MODE_BOOTLOADER = 3
-} FuCorsairDeviceMode;
-
-static gboolean
-fu_corsair_device_command(FuDevice *device, guint8 *data, guint timeout, GError **error)
-{
-	FuCorsairDevice *self = FU_CORSAIR_DEVICE(device);
-	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(device));
-	gsize actual_len = 0;
-	gboolean ret;
-
-	ret = g_usb_device_interrupt_transfer(usb_device,
-					      self->epout,
-					      data,
-					      self->cmd_write_size,
-					      &actual_len,
-					      timeout,
-					      NULL,
-					      error);
-	if (!ret) {
-		g_prefix_error(error, "failed to write command: ");
-		return FALSE;
-	}
-	if (actual_len != self->cmd_write_size) {
-		g_set_error(error,
-			    G_IO_ERROR,
-			    G_IO_ERROR_INVALID_DATA,
-			    "wrong size written: %" G_GSIZE_FORMAT,
-			    actual_len);
-		return FALSE;
-	}
-
-	memset(data, 0, CORSAIR_MAX_CMD_SIZE);
-
-	ret = g_usb_device_interrupt_transfer(usb_device,
-					      self->epin,
-					      data,
-					      self->cmd_read_size,
-					      &actual_len,
-					      timeout,
-					      NULL,
-					      error);
-	if (!ret) {
-		g_prefix_error(error, "failed to get command response: ");
-		return FALSE;
-	}
-	if (actual_len != self->cmd_read_size) {
-		g_set_error(error,
-			    G_IO_ERROR,
-			    G_IO_ERROR_INVALID_DATA,
-			    "wrong size read: %" G_GSIZE_FORMAT,
-			    actual_len);
-		return FALSE;
-	}
-
-	if (data[CORSAIR_OFFSET_CMD_STATUS] != 0) {
-		g_set_error(error,
-			    G_IO_ERROR,
-			    G_IO_ERROR_FAILED,
-			    "device replied with error: 0x%02x",
-			    data[CORSAIR_OFFSET_CMD_STATUS]);
-		return FALSE;
-	}
-
-	return TRUE;
-}
+void
+fu_corsair_device_finalize(GObject *object);
+gboolean
+fu_corsair_device_poll(FuDevice *device, GError **error);
```

Comment:
do you need this?

Reply:
it's for a case when mouse is turned off and receiver connected to the USB.  In this case the mouse can't be added because the plugin won't allow to add device without version (the version can be read only if a mouse is connected to the receiver). So I start polling until mouse is connected. Is there better way to do this?

Reply:
ohh i meant more the prototype.